p8e-api/src/main/kotlin/io/provenance/engine/grpc/v1/IndexGrpc.kt

```diff
@@ -132,9 +144,38 @@ class IndexGrpc(
     ) {
         P8eMDC.set(publicKey(), clear = true)
 
-        indexService.findLatestByScopeUuids(request.uuidsList.map { it.toUuidProv() })
+        val indexedScopes = indexService.findLatestByScopeUuids(request.uuidsList.map { it.toUuidProv() })
             .toScopeWrappers()
-            .complete(responseObserver)
+
+        log.debug("Fetched ${indexedScopes.scopesCount}/${request.uuidsCount} scopes from db")
+
+        var allScopes = indexedScopes
+
+        if (indexedScopes.scopesCount != request.uuidsCount) {
+            val fetchedUuids = indexedScopes.scopesList.map { it.scope.uuid.toUuidProv() }.toSet()
+
+
+            val chainScopes = request.uuidsList
+                .map { it.toUuidProv() }
+                .filterNot { fetchedUuids.contains(it) }
+                .also { log.debug("Attempting to fetch an additional ${it.count()} scopes from chain") }
+                .map {
```

Comment:
is this equivalent to `mapNotNull`?

Reply:
oh yeah, I think so, good call I'll change that